### ISSUE TITLE
set_visible_devices with MirroredStrategy causes NCCL warnings

### ISSUE THREAD

**Issue by Antymon** [tensorflow/tensorflow#44797]
**System information**
- I have written custom code
- OS Platform and Distribution: Linux Ubuntu 18.04
- TensorFlow installed `FROM tensorflow/tensorflow:2.3.0-gpu`
- TensorFlow version 2.3.0 
- Python version 3.6.9
- CUDA/cuDNN version: 10.1
- GPU model and memory: nvidia dgx, 8xV100, volta architecture, 8x16GB

**Describe the current behavior**

If `tf.config.set_visible_devices` is used presumably "Virtual GPUs" are created, which when combined with default `MirroredStrategy` leads to the following warning:

> WARNING:tensorflow:NCCL is not supported when using virtual GPUs, falling back to reduction to one device

I started setting the `CUDA_VISIBLE_DEVICES` env variable to go around the need for `set_visible_devices` but that is not ideal:
- it defeats the purpose of the `set_visible_devices`  
- it is not convenient
- my tensorboard profiler plugin started to crash with segfault after the `keras.fit` training session is over but before profiling data buffer is flushed (perhaps this is fair game as it seems that CUDA 10.2 is needed for that to officially work for multi-GPU)

**Describe the expected behavior**

I would like to use `set_visible_devices` without causing any problems listed above.

**Standalone code to reproduce the issue**

I will provide this at a later stage if necessary, but I believe just using a combination of 2 APIs above should reproduce the issue.

**Comment by amahendrakar** [comment]
@Antymon,
In order to expedite the trouble-shooting process, could you please provide a minimal code snippet to reproduce the issue reported here. Thanks!

**Comment by Antymon** [comment]
Assuming a multi-GPU node with CUDA 10.1 run docker container:
`docker run -it --rm tensorflow/tensorflow:2.3.0-gpu`

Create `limit_devices_tf.py`:
```
import os
import sys

import tensorflow as tf
from tensorflow.keras import layers, models

def set_visible_devices(device_idx: str, device_type: str, use_tf: bool):
    """
    Limit visibility of devices available in the cluster for the current application.

    @param device_idx: comma-delimited indexes or "ALL"
    @param device_type: [XLA_](CPU|GPU), e.g., "GPU"
    @param use_tf: set visibility of devices through tensorflow APIs or no

    @return: list of selected devices
    """

    if use_tf:
      filtered_type_devices = tf.config.list_physical_devices(device_type=device_type)
      visible_devices = select_devices_by_index(device_idx, filtered_type_devices)

      # always add CPU
      has_cpu = device_type[-3:] == "CPU"
      if not has_cpu:
          visible_devices += tf.config.list_physical_devices(
              device_type="{}CPU".format(device_type[:-3]).upper()
          )

      tf.config.set_visible_devices(visible_devices)
    elif device_type == "GPU":
      os.environ["CUDA_VISIBLE_DEVICES"]=device_idx
    else:
      raise NotImplementedError()

def select_devices_by_index(device_idx: str, devices: list):
    """
    @param device_idx: comma-delimited indexes or "ALL"
    @param devices: list of PhysicalDevices properties
    @return: subset of a list corresponding to indexes
    """
    assert len(devices) > 0
    filtered_devices = []
    if device_idx.upper() == "ALL":
        filtered_devices = devices
    else:
        device_idx = device_idx.split(",")
        assert len(devices) >= len(device_idx)
        for i in device_idx:
            for d in devices:
                if i == d.name.split(":")[-1]:
                    filtered_devices.append(d)

    print("{} devices selected:".format(len(filtered_devices)))
    for d in filtered_devices:
        print(d.name)

    return filtered_devices

def build_model():
    # build dummy model classification model
    model = models.Sequential()
    model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.Flatten())
    model.add(layers.Dense(64, activation='relu'))
    model.add(layers.Dense(10))
    model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
    return model

if __name__ == "__main__":
  use_tf = sys.argv[1].upper() == "TRUE"
  set_visible_devices(device_idx="0,1", device_type="GPU", use_tf = use_tf)
  strategy = tf.distribute.MirroredStrategy()
  with strategy.scope():
      model = build_model()
```
Run following in bash;
```
python limit_devices_tf.py true &> tf_true.txt
python limit_devices_tf.py false &> tf_false.txt
```
Only the former out of 2 output files is expected to contain:
> WARNING:tensorflow:NCCL is not supported when using virtual GPUs, fallingback to reduction to one device

**Comment by nikitamaia** [comment]
Hi @Antymon, I was able to reproduce this using 4 GPUs on GCP. This does seem odd that `set_visible_devices` should result in  virtual GPUs. As a workaround while this issue is being investigated further, have you tried passing in the devices argument to the strategy? eg `tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])`

**Comment by Antymon** [comment]
Hi, it's speculation. I didn't look inside the source to see what exactly is going on. 

As for the device parameter you pointed at, I had some other problems with it. I didn't know where I should take those names from. I ended up mapping filtered logical devices to their names (since names of physical ones caused problems), but there was something not right with it as well. I gave up and used `CUDA_VISIBLE_DEVICES` set as in example above. It does the main work, so I moved on.
